pep-0695.rst

```diff
@@ -809,30 +816,276 @@ It also adds an AST node type that represents a type parameter.
+recursive, and type variable bounds may refer back to the current class. If these
+expressions were evaluated eagerly, users would need to enclose such expressions in
+quotes to prevent runtime errors. :pep:`563` and :pep:`649` detail the problems with
+this situation for type annotations.
+
+To prevent a similar situation with the new syntax proposed in this PEP, we propose
+to use lazy evaluation for these expressions, similar to the approach in :pep:`649`.
+Specifically, each expression will be saved in a code object, and the code object
+is evaluated only when the corresponding attribute is accessed (``TypeVar.__bound__``,
+``TypeVar.__constraints__``, or ``TypeAlias.__value__``). After the value is
+successfully evaluated, the value is saved and later calls will return the same value
+without re-evaluating the code object.
+
+If :pep:`649` is implemented, additional evaluation mechanisms should be added to
+mirror the options that PEP provides for annotations. In the current version of the
+PEP, that might include adding an ``__evaluate_bound__`` method to ``TypeVar`` taking
+a ``format`` parameter with the same meaning as in PEP 649's ``__annotate__`` method
+(and a similar ``__evaluate_constraints__`` method, as well as an ``__evaluate_value__``
+method on ``TypeAliasType``).
+However, until PEP 649 is accepted and implemented, only the default evaluation format
+(PEP 649's "VALUE" format) will be supported.
+
+As a consequence of lazy evaluation, the value observed for an attribute may
+depend on the time the attribute is accessed.
+
+::
+
+   X = int
+
+   class Foo[T: X, U: X]:
+       t, u = T, U
+
+   print(Foo.t.__bound__)  # prints "int"
+   X = str
+   print(Foo.u.__bound__)  # prints "str"
+
+Similar examples affecting type annotations can be constructed using the
+semantics of PEP 563 or PEP 649.
+
+A naive implementation of lazy evaluation would handle class namespaces
+incorrectly, because functions within a class do not normally have access to
+the enclosing class namespace. The implementation will retain a reference to
+the class namespace so that class-scoped names are resolved correctly.
+
+.. _695-scoping-behavior:
+
+Scoping Behavior
+----------------
+
+The new syntax requires a new kind of scope that behaves differently
+from existing scopes in Python. Thus, the new syntax cannot be described exactly in terms of
+existing Python scoping behavior. This section specifies these scopes
+further by reference to existing scoping behavior: the new scopes behave
+like function scopes, except for a number of minor differences listed below.
+
+All examples include functions introduced with the pseudo-keyword ``def695``.
+This keyword will not exist in the actual language; it is used to
+clarify that the new scopes are for the most part like function scopes.
+
+``def695`` scopes differ from regular function scopes in the following ways:
+
+- If a ``def965`` scope is immediately within a class scope, or within another
```

Comment:
drive by
```suggestion
- If a ``def695`` scope is immediately within a class scope, or within another
```

Reply:
Good thing this is not actually becoming a keyword :)